src/contracts.nit

```diff
@@ -0,0 +1,885 @@
+end
+
+redef class APropdef
+	redef fun check_callsite(v)
+	do
+		v.visited_method = self
+	end
+end
+
+redef class AMethPropdef
+
+	# Execute all method verification scope flow and typing.
+	# It also execute an ast validation to define all parents and all locations
+	private fun do_all(toolcontext: ToolContext)
+	do
+		self.validate
+		# FIXME: The `do_` usage it is maybe to much (verification...). Solution: Cut the `do_` methods into simpler parts
+		self.do_scope(toolcontext)
+		self.do_flow(toolcontext)
+		self.do_typing(toolcontext.modelbuilder)
+	end
+
+	# Entry point to create a contract (verification of inheritance, or new contract).
+	redef fun create_contracts(v)
+	do
+		v.ast_builder.check_mmodule(mpropdef.mclassdef.mmodule)
+
+		v.current_location = self.location
+		v.is_intro_contract = mpropdef.is_intro
+
+		if n_annotations != null then
+			for n_annotation in n_annotations.n_items do
+				check_annotation(v,n_annotation)
+			end
+		end
+
+		if not mpropdef.is_intro and not v.find_no_contract then
+			self.check_redef(v)
+		end
+
+		# reset the flag
+		v.find_no_contract = false
+	end
+
+	# Verification of the annotation to know if it is a contract annotation.
+	# If this is the case, we built the appropriate contract.
+	private fun check_annotation(v: ContractsVisitor, n_annotation: AAnnotation)
+	do
+		if n_annotation.name == "expects" then
+			var exist_contract = mpropdef.mproperty.check_exist_expect
+			mpropdef.construct_contract(v, self.n_signature.as(not null), n_annotation, mpropdef.mproperty.mexpect.as(not null), exist_contract)
+		else if n_annotation.name == "ensures" then
+			var exist_contract = mpropdef.mproperty.check_exist_ensure
+			mpropdef.construct_contract(v, self.n_signature.as(not null), n_annotation, mpropdef.mproperty.mensure.as(not null), exist_contract)
+		else if n_annotation.name == "no_contract" then
+			# no_contract found set the flag to true
+			v.find_no_contract = true
+		end
+	end
+
+	# Verification if the method have an inherited contract to apply it.
+	private fun check_redef(v: ContractsVisitor)
+	do
+		# Check if the method has an attached contract
+		if not mpropdef.has_contract then
+			if mpropdef.mproperty.intro.has_contract then
+				mpropdef.has_contract = true
+			end
+		end
+	end
+
+	# Adapt the block to use the contracts
+	private fun adapt_block_to_contract(v: ContractsVisitor, contract: MContract, n_mpropdef: AMethPropdef)
+	do
+		contract.adapt_block_to_contract(v, n_mpropdef)
+		mpropdef.has_contract = true
+		n_mpropdef.do_all(v.toolcontext)
+	end
+end
+
+redef class ASelfExpr
+	# Defined the self type
+	fun set_mtype(mtype: nullable MType)
+	do
+		self.mtype = mtype
+	end
+end
+
+redef class MSignature
+	# Adapt signature for a expect condition
+	# Removed the return type is it not necessary
+	private fun adapt_to_condition(return_mtype: nullable MType): MSignature do
+		var adapt_msignature = self.copy
+		adapt_msignature.set_return(return_mtype)
+		return adapt_msignature
```

Comment:
just do it, no need ad hoc helpers:

`return new MSignature(self.mparameters.to_a, return_mtype)`